File: 52-n-queens-ii/52-n-queens-ii.py
class Solution:
    def totalNQueens(self, n: int) -> int:
        res = 0
        board = [["." for _ in range(n)] for _ in range(n)]

        rows = set()
        diagUp = set()
        diagDown = set()

        def recSearch(j: int) -> None:
            nonlocal res, board
            if j == n:
                res += 1
                return
            for i in range(n):
                if not((i in rows) or ((i + j) in diagUp) or ((i - j) in diagDown)):
                    # try this pos and add a queen
                    rows.add(i)
                    diagUp.add(i + j)
                    diagDown.add(i - j)
                    board[i][j] = "Q"
                    recSearch(j + 1)
                    rows.discard(i)
                    diagUp.discard(i + j)
                    diagDown.discard(i - j)
                    board[i][j] = "."  # remove queen after
        recSearch(0)
        return res

File: 52-n-queens-ii/52-n-queens-ii.cpp
class Solution {
    
public:
    int res = 0;
    unordered_set<int> rows;
    unordered_set<int> diagUp;
    unordered_set<int> diagDown;
    int totalNQueens(int n) {
        recSearch(0, n);
        return res;
    }
    void recSearch(int j, int n) {
        if (j == n) {
            res++;
            return;
        }
        for (int i=0; i<n; i++) {
            if (not((rows.find(i) != rows.end()) or (diagUp.find(i + j) != diagUp.end()) or (diagDown.find(i - j) != diagDown.end()))) {
                // try this pos and add a queen
                rows.insert(i);
                diagUp.insert(i + j);
                diagDown.insert(i - j);
                recSearch(j + 1, n);
                rows.erase(i);
                diagUp.erase(i + j);
                diagDown.erase(i - j);
            }
        }
    }
};